conda-store-server/conda_store_server/worker/tasks.py

```diff
@@ -62,21 +63,87 @@ def task_update_storage_metrics(self):
     )
 
 
+"""
+Pierre - May 29th 2022
+This is a different version of task_update_conda_channels.
+It's designed to run one task per channel, with a lock to avoid triggering twice a task for a given channel.
+Two scenarios :
+- either you're using Redis (that should be the case if you have multiple celery workers) : then the lock is handled by Redis.
+- either you don't (that's fine if you're using a single celery worker) : then the lock is a Filelock
+
+The reason behind having the task running only once at a time is
+to avoid integrity exceptions when running channel.update_packages.
+
+Sources :
+Lock : http://loose-bits.com/2010/10/distributed-task-locking-in-celery.html
+Redis : https://pypi.org/project/redis/
+https://stackoverflow.com/questions/12003221/celery-task-schedule-ensuring-a-task-is-only-executed-one-at-a-time
+
+"""
+
+
 @current_app.task(base=WorkerTask, name="task_update_conda_channels", bind=True)
 def task_update_conda_channels(self):
     conda_store = self.worker.conda_store
+
+    conda_store.ensure_conda_channels()
+    for channel in api.list_conda_channels(conda_store.db):
+        send_task("task_update_conda_channel", args=[channel.name], kwargs={})
+
+
+@current_app.task(base=WorkerTask, name="task_update_conda_channel", bind=True)
+def task_update_conda_channel(self, channel_name):
+
+    conda_store = self.worker.conda_store
+
+    # sanitize the channel name as it's an URL, and it's used for the lock.
+    sanitizing = {
+        "https": "",
+        "http": "",
+        ":": "",
+        "/": "_",
+        "?": "",
+        "&": "_",
+        "=": "_",
+    }
+    channel_name_sanitized = channel_name
+    for k, v in sanitizing.items():
+        channel_name_sanitized = channel_name_sanitized.replace(k, v)
+
+    task_key = f"lock_{self.name}_{channel_name_sanitized}"
+
+    is_locked = False
+
+    if conda_store.redis_url is not None:
```

Comment:
An alternative to this pattern and using try/finally is using the https://docs.python.org/3/library/contextlib.html#contextlib.ExitStack. This way you don't have to enter/exit manually.